### ISSUE TITLE
Regarding File Uploads Vulnerabilities

### ISSUE THREAD

**Issue by preetkaran20** [zaproxy/zaproxy#6154]
**Is your feature request related to a problem? Please describe.**
Currently I think ZAP doesn't have a way to check for File Upload vulnerabilities like say file name causing XSS or Path traversal or other vulnerabilities associated with File Uploads. As discussed with @psiinon this is not a trivial task as once file is uploaded ZAP needs to find the uploaded files Url for complete testing of this vulnerability.

**Describe the solution you'd like**
Upload the file and check for url's in response or may be spider the application again or ask for the uploaded files url location in active scanner options and try to find the same file names. Sometimes file names are changed by the server and in those cases it might be little tricky so in the start we can just search in response and later we can think or other things.

**Describe alternatives you've considered**

**Additional context**
I was including a new vulnerability in @Sasanlabs/VulnerableApp and didn't find any way to scan this vulnerability using ZAP.

**Would you like to help implementing this feature?**
Yes

**Comment by thc202** [comment]
How was the file uploaded?

**Comment by preetkaran20** [comment]
I uploaded the file using Ajax call using formData

**Comment by thc202** [comment]
That should be covered by Multipart Form-Data variant (`VariantMultipartFormParameters`), wasn't the file name field tested?

**Comment by psiinon** [comment]
Interesting. I guess in theory the persistant XSS rule could find them. But I'm pretty sure it wont attempt to upload a file with a name as the payload. That would be a nice enhancement...

**Comment by thc202** [comment]
The file name field is one of the parameters.

**Comment by preetkaran20** [comment]
`VariantMultipartFormParameters` file is extracting the parameters correctly. May be active scanners are not fuzzing these properties.


![image](https://user-images.githubusercontent.com/13570884/91873837-239e1180-ec97-11ea-9f47-2a82394808ea.png)

**Comment by thc202** [comment]
The scan rules don't usually check the parameter type, most likely they are just not finding the issue.

**Comment by preetkaran20** [comment]
Hi @thc202 @kingthorin,

please assign this enhancement to me.

thanks,
Karan

**Comment by preetkaran20** [comment]
Hi @thc202 @psiinon @kingthorin ,

I was debugging the same for finding reflected XSS and I found that ZAP can find reflect XSS just the problem was the context in Html. Till now i was returning Json response something like:
```
{
  "content": "upload/-1800538472_fuzzer-view.png",
  "isValid": true
}
```
and because the context was not Html hence it was not reporting it. Once i changed the context to html something like:
```
<img src="fuzzer-view.png"/>
```
ZAP was able to find the vulnerability.

For reflected XSS i think we need not to do change anything. However i am having one doubt that in general now applications are doing Ajax calls and then appending the result to the dom tree so say the returned json could cause XSS but because rules are always searching for Html context hence they are not able to found the vulnerability.
Is there any solution we have for such problem ? or have we done something to handle these usecases.

thanks,
Karan

**Comment by preetkaran20** [comment]
As @thc202  has mentioned, need to look at following addons:
https://github.com/zaproxy/zap-extensions/tree/master/addOns/frontendscanner
https://github.com/zaproxy/front-end-tracker

**Comment by preetkaran20** [comment]
Hi @thc202 @kingthorin ,

I was trying to debug the PersistentXSS Scan rule but i am not able to understand how does this works. @psiinon told me that it will spider the application after injection but some how i am not able to find the same thing. there is some code present in PersistentXssUtils for finding the sink but i am not sure how is that populated. While debugging i found that it is not getting populated and hence the persistentXSS scan rule doesn't do anything.

Files i am looking at are: PersistentXssScanRule, PersistentXssSpiderScanRule.
Am i missing some configuration ? Please correct me. 

thanks,
Karan

**Comment by kingthorin** [comment]
The existing persistent XSS scan rule is made up of 3 components as you can see via the scan policy or code base.

The persistent XSS scan rule is subject to a [known issue](https://github.com/zaproxy/zaproxy/issues/4656) where only the last param in a request is effectively tested. Theres also a rather large [PR](https://github.com/zaproxy/zap-extensions/pull/2443) in need of review/testing to fix it up.

**Comment by preetkaran20** [comment]
https://github.com/portswigger/upload-scanner

**Comment by preetkaran20** [comment]
PR: https://github.com/SasanLabs/owasp-zap-fileupload-addon/pull/1

**Comment by preetkaran20** [comment]
Closing this issue as https://github.com/SasanLabs/owasp-zap-fileupload-addon addon is released for the dev version and will be available once 2.11.0 is available.

thanks,
Karan
